samples/simple_txrx/src/main.c

```diff
@@ -0,0 +1,290 @@
+#include <zephyr/shell/shell.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+
+#include "rail.h"
+#include "rail_config.h"
+#include "pa_conversions_efr32.h"
+
+LOG_MODULE_REGISTER(app);
+
+#ifdef RAIL0_CHANNEL_GROUP_1_PROFILE_WISUN_OFDM
+#  if !defined(HARDWARE_BOARD_HAS_EFF)
+     BUILD_ASSERT(SL_RAIL_UTIL_PA_SELECTION_SUBGHZ == RAIL_TX_POWER_MODE_OFDM_PA,
+		  "Please use the OFDM PA settings in the sl_rail_util_pa_config.h for OFDM phys");
+#  endif
+#  if defined(HARDWARE_BOARD_HAS_EFF) && RAIL_SUPPORTS_EFF
+     BUILD_ASSERT(SL_RAIL_UTIL_PA_SELECTION_SUBGHZ >= RAIL_TX_POWER_MODE_OFDM_PA_EFF_30DBM,
+		  "Please use the OFDM PA for EFF settings in the sl_rail_util_pa_config.h for OFDM phys");
+#  endif
+#endif
+
+static const struct gpio_dt_spec sw0 = GPIO_DT_SPEC_GET(DT_ALIAS(sw0), gpios);
+static const struct gpio_dt_spec led_rx = GPIO_DT_SPEC_GET(DT_ALIAS(led0), gpios);
+#if DT_NODE_EXISTS(DT_ALIAS(led1))
+static const struct gpio_dt_spec led_tx = GPIO_DT_SPEC_GET(DT_ALIAS(led1), gpios);
+#else
+static const struct gpio_dt_spec led_tx = led_rx;
+#endif
+
+enum {
+	EV_RAIL_RX     = BIT(0),
+	EV_BTN_PRESSED = BIT(1),
+};
+
+struct {
+	RAIL_Handle_t rail_handle;
+	struct k_event rx_event;
+	int channel;
+	const uint8_t *payload;
+	int payload_len;
+} app_ctx;
+
+void rx_packets(RAIL_Handle_t rail_handle)
+{
+	uint8_t rx_frame[32];
+	RAIL_RxPacketHandle_t handle;
+	RAIL_RxPacketInfo_t info;
+	RAIL_Status_t status;
+
+	for (;;) {
+		handle = RAIL_GetRxPacketInfo(rail_handle, RAIL_RX_PACKET_HANDLE_OLDEST_COMPLETE,
+					      &info);
+		if (handle == RAIL_RX_PACKET_HANDLE_INVALID) {
+			return;
+		}
+		if (info.packetBytes < sizeof(rx_frame)) {
+			RAIL_CopyRxPacket(rx_frame, &info);
+		}
+		status = RAIL_ReleaseRxPacket(rail_handle, handle);
+		if (status) {
+			LOG_ERR("RAIL_ReleaseRxPacket(): %d", status);
+		}
+		if (info.packetBytes < sizeof(rx_frame)) {
+			LOG_HEXDUMP_INF(rx_frame, info.packetBytes, "rx data:");
+		} else {
+			LOG_INF("rx: skip large packet");
+		}
+		gpio_pin_toggle_dt(&led_rx);
+	}
+}
+
+void tx_packet(RAIL_Handle_t rail_handle, int channel, const uint8_t *payload, int len)
+{
+	RAIL_Status_t status;
+	int ret;
+
+	ret = RAIL_WriteTxFifo(rail_handle, payload, len, true);
+	if (ret != len) {
+		LOG_ERR("RAIL_WriteTxFifo(): %d", ret);
+		return;
+	}
+	gpio_pin_toggle_dt(&led_tx);
+	status = RAIL_StartTx(rail_handle, channel, RAIL_TX_OPTIONS_DEFAULT, NULL);
+	if (status) {
+		LOG_ERR("RAIL_StartTx(): %d ", status);
+	}
+	LOG_HEXDUMP_INF(payload, len, "tx data:");
+}
+
+void btn_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	/* This function is called from an ISR context. So, transfer the real
+	 * processing to the main loop.
+	 */
+	k_event_post(&app_ctx.rx_event, EV_BTN_PRESSED);
+}
+
+void cli_send(const struct shell *sh, size_t argc, char **argv)
+{
+	tx_packet(app_ctx.rail_handle, app_ctx.channel, app_ctx.payload, app_ctx.payload_len);
```

Comment:
I don't know if cli is another thread or ISR, but I would suggest to the same style as button for easier maintain

Reply:
`cli_send()` is called from another thread. In a shell, the user expect to run synchronous actions.

Reply:
If it's an another thread wouldn't be a MUTEX a good idea? I know the possibility of pushing the button and cli is low

Reply:
Do you think we need to protect rx and tx wit the same mutex or we can use two independent mutexes?

Reply:
I think one will be enough, as only TX could be initiated from a different thread.

Reply:
In fact, cli thread can send a packet while the main thread is receiving one.

Reply:
That won't be an issue as it is modifying different part of the RAIL, and after send it will automatically go back to receive.